nginxinc/kubernetes-ingress
# Revert to original main-template without pod downtime

### Proposed changes

This PR introduces the following changes:

- when a user deletes custom `main-template` from the ConfigMap obj, NIC will revert to the original, default template
- there is no need to restart NIC pod and introduce downtime anymore


### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc


### Testing:

- deploy NIC and verify initial ConfigMap:
```shell
➜  common git:(fix/github-5906) ✗ k describe configmaps -n nginx-ingress nginx-config
Name:         nginx-config
Namespace:    nginx-ingress
Labels:       <none>
Annotations:  <none>

Data
====

BinaryData
====

Events:
  Type    Reason   Age                From                      Message
  ----    ------   ----               ----                      -------
  Normal  Updated  42m (x9 over 20h)  nginx-ingress-controller  Configuration from nginx-ingress/nginx-config was updated
  Normal  Updated  31m (x3 over 35m)  nginx-ingress-controller  Configuration from nginx-ingress/nginx-config was updated
```
- check content of `/etc/nginx/nginx.conf` in the NIC pod
```shell
nginx@nginx-ingress-68449896cb-4jjpp:/etc/nginx$ head nginx.conf
worker_processes  auto;

daemon off;

error_log  stderr notice;
pid        /var/lib/nginx/nginx.pid;
load_module modules/ngx_fips_check_module.so;

load_module modules/ngx_http_js_module.so;
```
- apply custom `main-template` using ConfigMap
beginning of the custom `main-template`
```shell
➜  common git:(fix/github-5906) ✗ head nginx-config-main-template-02.yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: nginx-config
  namespace: nginx-ingress
data:
  main-template: |
    # VERSION 2
    # main template from config map
    # this is a test
```
- verify the main template is applied
```shell
I20241015 12:15:56.435890   1 verify.go:90] success, version 4 ensured. took: 161.217208ms
I20241015 12:15:56.436318   1 controller.go:262] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"nginx-ingress", Name:"nginx-config", UID:"ae356055-5493-47fc-b107-5d16e7a71c10", APIVersion:"v1", ResourceVersion:"185894", FieldPath:""}): type: 'Normal' reason: 'Updated' Configuration from nginx-ingress/nginx-config was updated
2024/10/15 12:15:56 [notice] 78#78: exit
```
```shell
nginx@nginx-ingress-68449896cb-4jjpp:/etc/nginx$ head nginx.conf
# VERSION 2
# main template from config map
# this is a test
#
worker_processes  auto;
daemon off;
error_log  stderr notice;
pid        /var/lib/nginx/nginx.pid;
load_module modules/ngx_fips_check_module.so;
load_module modules/ngx_http_js_module.so;
```
```shell
➜  common git:(fix/github-5906) ✗ k describe configmaps -n nginx-ingress nginx-config
Name:         nginx-config
Namespace:    nginx-ingress
Labels:       <none>
Annotations:  <none>

Data
====
main-template:
----
# VERSION 2
# main template from config map
# this is a test
#
worker_processes  auto;
daemon off;
error_log  stderr notice;
pid        /var/lib/nginx/nginx.pid;
load_module modules/ngx_fips_check_module.so;
load_module modules/ngx_http_js_module.so;

events {
    worker_connections  1024;
}
. . . 
```

- remove the custom `main-template` from the ConfigMap obj (apply an empty ConfigMap)
```shell
➜  common git:(fix/github-5906) ✗ kaf nginx-config.yaml
configmap/nginx-config configured
```
```shell
➜  common git:(fix/github-5906) ✗ k describe configmaps -n nginx-ingress nginx-config
Name:         nginx-config
Namespace:    nginx-ingress
Labels:       <none>
Annotations:  <none>

Data
====

BinaryData
====

Events:
  Type    Reason   Age                From                      Message
  ----    ------   ----               ----                      -------
  Normal  Updated  51m (x9 over 21h)  nginx-ingress-controller  Configuration from nginx-ingress/nginx-config was updated
  Normal  Updated  60s (x5 over 44m)  nginx-ingress-controller  Configuration from nginx-ingress/nginx-config was updated
```
- verify `/etc/nginx/nginx.conf` is reverted back to the original config file
```shell
nginx@nginx-ingress-68449896cb-4jjpp:/etc/nginx$ head nginx.conf
worker_processes  auto;

daemon off;

error_log  stderr notice;
pid        /var/lib/nginx/nginx.pid;
load_module modules/ngx_fips_check_module.so;

load_module modules/ngx_http_js_module.so;
```
